blog/2024-01-12-implement-async-storage-data-streaming-pytorch/index.md

```diff
@@ -0,0 +1,256 @@
+        cv2.circle(img, (int(annotation['mouth-x']), int(annotation['mouth-y'])), 12, (100, 0, 0), 1)
+
+        # Drawing ears
+        for ear_side in ['right', 'left']:
+            pts = np.array([[annotation[f'{ear_side}-ear-{i}-x'], annotation[f'{ear_side}-ear-{i}-y']] for i in range(1, 4)], np.int32)
+            pts = pts.reshape((-1, 1, 2))
+            cv2.polylines(img, [pts], True, (0, 255, 255))
+
+```
+
+The `process_record` method prepares the images. It decodes the byte stream of each image and annotates imagegs (for the sake of the example) using metadata provided with each record, such as bounding box coordinates for features of interest like the eyes or mouth in the 'cats' dataset.
+
+With this configuration, the `RemoteCatsDataset` is ready for integration with PyTorch's DataLoader, which can operate multiple processes in parallel using several workers. For example
+
+```python
+from torch.utils.data import DataLoader
+
+client = Client(HOST, api_token=API_TOKEN)
+
+# Create an instance of your dataset
+remote_dataset = BufferedShuffleIterableDataset(
+    client, BUCKET, DATASET, start=1, stop=1000
+)
+
+# Create a DataLoader
+data_loader = DataLoader(remote_dataset, batch_size=100, num_workers=8)
+
+# Display each image
+for images in data_loader:
+    # You would oerform your training steps here
+    # Example: train_step(model, images)
+    for img in images:
+        plt.imshow(img)
+        plt.show()
+```
+
+![Cat number 1](./img/cat-number-1.jpeg)
+
+![Cat number 2](./img/cat-number-2.jpeg)
+
+![Cat number 3](./img/cat-number-3.jpeg)
+
+![Cat number 4](./img/cat-number-4.jpeg)
+
+Using the `DataLoader` with multiple workers allows for efficient data loading and preprocessing. This parallelism is especially beneficial when training deep-learning models on large datasets where I/O bound operations could become a bottleneck.
+
+## Implementing a Buffered Shuffle Iterable Dataset
+Normally, the entire training set is shuffled after each epoch to avoid over fitting. However, in our scenario, we don't have the full data available on disk, and we don't have all the indexes of the pictures in the database neither. We provide a start and stop index to the database, and it returns the data in that range. So, we can't shuffle the data in the database, and we can't shuffle the data in memory either.
+
+Implementing a Buffered Shuffle Iterable Dataset in PyTorch is a trade-off to be able to shuffle the data we fetch from the remote server. This works, but we have to be careful not to get duplicates when using several workers, and we need to define a different data stream for each seperate process. 
+
+One way to handle this, is to look at the worker information for each separate process and request different data ranges from the database.
+
+```python
+import random
+import torch
+
+class BufferedShuffleIterableDataset(RemoteCatsDataset):
```

Comment:
This is really interesting. I could implement it on the storage side. A user could use something like this:

```python
bucket.query(self.dataset_name, start=worker_start, stop=worker_stop, random=true)
```

Reply:
That would be great! We should definitely keep that in backlog.